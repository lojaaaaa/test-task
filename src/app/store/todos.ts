import { getLocalstorage, setLocalstorage } from 'src/shared/lib';
import { create } from 'zustand'

export interface ITodo {
  id: string;
  value: string;
  isCompleted: boolean;
}

type State = {
  todos: ITodo[];
}

type Action = {
  createTodo: (todo: ITodo) => void;
  updateTodos: (todos: ITodo[]) => void;
  clearCompletedTodos: () => void;
  setTodoStatus: (id: string, isCompleted: boolean) => void;
}

export const useTodosStore = create<State & Action>((set) => ({
  todos: getLocalstorage(),

  createTodo: (todo) => set((state) => {
    const updatedTodos = [...state.todos, todo];
    setLocalstorage(updatedTodos);
    return { todos: updatedTodos };
  }),

  updateTodos: (updatedTodos) => set(() => {
    setLocalstorage(updatedTodos);
    return { todos: updatedTodos };
  }),

  clearCompletedTodos: () => set((state) => {
    const updatedTodos = state.todos.filter(todo => !todo.isCompleted);
    setLocalstorage(updatedTodos);
    return { todos: updatedTodos };
  }),

  setTodoStatus: (id, isCompleted) => set((state) => {
    const updatedTodos = state.todos.map(todo =>
      todo.id === id ? { ...todo, isCompleted } : todo
    );
    setLocalstorage(updatedTodos);
    return { todos: updatedTodos };
  }),
}));
