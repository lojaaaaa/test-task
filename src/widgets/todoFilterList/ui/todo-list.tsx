import { filterOptions } from "../fixtures";

interface TodoFilterListProps {
  activeKeys: string[];
  toggleActiveKey: (key: string) => void;
};

export const TodoFilterList = ({ activeKeys, toggleActiveKey }: TodoFilterListProps) => (
  <ul className="flex gap-4">
    {filterOptions.map(({ key, label }) => (
      <li
        key={key}
        onClick={() => toggleActiveKey(key)}
        className={activeKeys.includes(key) ? 'cursor-pointer rounded-lg p-2 border border-red-200' : 'cursor-pointer p-2'}
      >
        {label}
      </li>
    ))}
  </ul>
);
