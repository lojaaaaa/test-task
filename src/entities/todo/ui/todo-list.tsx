
import { ITodo } from 'src/app/store';
import { TodoItem } from 'src/entities/todo';

interface TodoListProps {
  items: ITodo[];
}

export const TodoList = ({ items }: TodoListProps) => (
  <ul className="flex flex-col">
    {
      items.length
        ? items.map(({ id, value, isCompleted }) => (
          <TodoItem key={id} id={id} value={value} isCompleted={isCompleted} />
        ))
        : <span className="text-gray-500 text-center px-2 py-4">None</span>
    }
  </ul>
);
