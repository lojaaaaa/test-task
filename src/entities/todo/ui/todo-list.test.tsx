import { render, screen } from '@testing-library/react';
import { TodoList } from './todo-list';
import { ITodo } from 'src/app/store/todos';

describe('TodoList Component', () => {
  const todos: ITodo[] = [
    { id: '1', value: 'First task', isCompleted: false },
    { id: '2', value: 'Second task', isCompleted: true },
  ];

  it('renders a list of todos when items are provided', () => {
    render(<TodoList items={todos} />);

    const firstTask = screen.getByText('First task');
    const secondTask = screen.getByText('Second task');

    expect(firstTask).toBeInTheDocument();
    expect(secondTask).toBeInTheDocument();
  });

  it('renders "None" when the todo list is empty', () => {
    render(<TodoList items={[]} />);

    const noneElement = screen.getByText('None');

    expect(noneElement).toBeInTheDocument();
  });

  it('displays a line-through style for completed tasks', () => {
    render(<TodoList items={todos} />);

    const completedTask = screen.getByText('Second task');

    expect(completedTask).toHaveClass('line-through');
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <TodoList items={todos} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

});
