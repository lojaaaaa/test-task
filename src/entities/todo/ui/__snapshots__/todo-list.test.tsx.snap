// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList Component > matches the snapshot 1`] = `
<DocumentFragment>
  <ul
    class="flex flex-col"
  >
    <li
      class="px-2 py-4 border-b border-gray-200"
    >
      <label
        class="ant-checkbox-wrapper text-lg css-dev-only-do-not-override-1c0na6j"
      >
        <span
          class="ant-checkbox ant-wave-target css-dev-only-do-not-override-1c0na6j"
        >
          <input
            class="ant-checkbox-input"
            type="checkbox"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          First task
        </span>
      </label>
    </li>
    <li
      class="px-2 py-4 border-b border-gray-200"
    >
      <label
        class="ant-checkbox-wrapper ant-checkbox-wrapper-checked text-lg css-dev-only-do-not-override-1c0na6j"
      >
        <span
          class="ant-checkbox ant-wave-target css-dev-only-do-not-override-1c0na6j ant-checkbox-checked"
        >
          <input
            checked=""
            class="ant-checkbox-input"
            type="checkbox"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          <span
            class="line-through"
          >
            Second task
          </span>
        </span>
      </label>
    </li>
  </ul>
</DocumentFragment>
`;
